### PR TITLE
Parse kind and apiVersion anywhere in the document

### PR DESCRIPTION
**Pull request by drewgonzales360**
## Description

I ran into https://github.com/helm/helm-mapkubeapis/issues/46 while trying to use this plugin. In my helm manifest, I had

```yaml
apiVersion: scheduling.k8s.io/v1beta1
description: "really cool description and totally unpredictable"
kind: PriorityClass
```

so the `strings.Contains` bit wasn't working well for me.

I updated the `Map.yaml` structure to give a better reflection of the APIs that are changing. Instead of doing a string search through the file, I specifically parse the `apiVersion` and the `kind`, then specifically compare those two fields.

```yaml
- deprecatedAPI:
    apiVersion: extensions/v1beta1
    kind: Deployment
  newAPI:
    apiVersion: apps/v1
    kind: Deployment
  deprecatedInVersion: "v1.9"
  removedInVersion: "v1.16"
```

- Also added tests. Added a target in the makefile and added it to CI
- Refactored `ReplaceManifestUnSupportedAPIs` so it was easier to test.

### CODE REVIEW

File: pkg/v3/release.go [hunk 1]
```diff
@@ -100,3 +111,15 @@ func getLatestRelease(releaseName string, cfg *action.Configuration) (*release.R
 func getReleaseVersionName(rel *release.Release) string {
 	return fmt.Sprintf("%s.v%d", rel.Name, rel.Version)
 }
+
+func getKubernetesServerVersion(kubeConfig common.KubeConfig) (string, error) {
```
Comment (drewgonzales360) [inline pkg/v3/release.go:115]: Moved from common.go

File: pkg/v3/release.go [hunk 2]
```diff
@@ -87,7 +98,7 @@ func updateRelease(origRelease *release.Release, modifiedManifest string, cfg *a
 	newRelease.Info.Status = release.StatusDeployed
 	log.Printf("Add release version '%s' with updated supported APIs.\n", getReleaseVersionName(origRelease))
 	if err := cfg.Releases.Create(newRelease); err != nil {
-		return errors.Wrapf(err, "failed to create new release version '%s': %s", getReleaseVersionName(origRelease))
+		return errors.Wrapf(err, "failed to create new release version '%s'", getReleaseVersionName(origRelease))
```
Comment (drewgonzales360) [inline pkg/v3/release.go:101]: Static analyzer was throwing me an error because there was an extra `%s`

File: pkg/common/manifest.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		log.Printf("The resource %s/%s does not require mapping as the "+
+			"API is not deprecated or removed in Kubernetes '%s'\n",
+			mapping.DeprecatedAPI.APIVersion,
+			mapping.DeprecatedAPI.Kind,
+			kubeVersionStr,
+		)
+		return false, nil
+	}
+
+	if manifest.APIVersion == mapping.DeprecatedAPI.APIVersion && manifest.Kind == mapping.DeprecatedAPI.Kind {
```
Comment (drewgonzales360) [inline pkg/common/manifest.go:106]: Here is where I actually update the values in the document.

**Comment by drewgonzales360** [comment]
@hickeyma Can you take a look at this whenever you get a free moment, or assign it to someone else? 🙏

**Comment by drewgonzales360** [comment]
Hi @pgier! I just thought I'd ask if you knew anyone that could take a look at this? I just saw that you have the second most number of commits here 😅. Lemme know if there's anything I can do to make this easier to review!

**Review by pgier — COMMENTED** [review]
@drewgonzales360 sorry I don't have merge access to this repo, @hickeyma will have to do that.  In general, I like the idea of parsing the APIVersion and Kind instead of relying on a specific order, however my main concern with this is that it will make changes to the manifest other than just the APIVersion and Kind.  For example, the ordering would likely not be preserved, and I'm not sure if comments would be preserved.

Since you are already breaking up the manifest by individual documents/objects, what if instead of unmarshalling the objects, you just find and replace in each object string, and then append them back together at the end?

**Comment by drewgonzales360** [comment]
Hm, you're right about ordering and comments. 

The helm templates will only have [comments if installed](https://helm.sh/docs/chart_best_practices/templates/) with `--debug` 
> The comment above is visible when the user runs helm install --debug, while comments specified in {{- /* */}} sections are not.

So if people did that, we'd wipe out comments.

Ordering is also not handled. Although in the current functionality, this plugin only works when `apiVersion` and `kind` are in a specific order. For example, the plugin would not remap the following.

```yaml
kind: PriorityClass
apiVersion: scheduling.k8s.io/v1beta1
```

I think the trade-off is worth it that sense. The plugin would still work in more cases than it does now. I guess the question is, "is the data in the configmap meant to be human readable, and to what extent?". All the fields would be there, just in a different order, and the comments would only be missing if the chart was installed with `--debug`.

File: pkg/v3/release.go (hunk 1 above)
Comment (drewgonzales360) [reply pkg/v3/release.go:115]: It doesn't need to be 😄. There was only one call to it and it was in the same file. When I moved the call to it, I would have either had to export the function, or move the whole function. I decided to move it, but I can move it back. Whatever works

**Review by hickeyma — COMMENTED** [review]
@drewgonzales360 Thanks for pushing the PR. I like your changes in principle but as a simple workaround tool, I'd prefer to leave the implementation as is. I would hope that the tool would eventually become obsolete due to more stability in APIs, charts being updated in time or a better solution.

That is why I don't think it a solution that we will merge at this time.

**Comment by drewgonzales360** [comment]
Gotcha, sounds good. I'll close this one then.

**Comment by ericyew** [comment]
> @drewgonzales360 Thanks for pushing the PR. I like your changes in principle but as a simple workaround tool, I'd prefer to leave the implementation as is. I would hope that the tool would eventually become obsolete due to more stability in APIs, charts being updated in time or a better solution.
> 
> That is why I don't think it a solution that we will merge at this time.

This should be reconsidered as the plugin as is misses way too many deprecated api due to the manifest containing random permutation of \n and \r. Even if the tool could at least call out the deprecated api that are detected so that we don't need to always manually double check each cluster.

**Comment by ericyew** [comment]
> @drewgonzales360 Thanks for pushing the PR. I like your changes in principle but as a simple workaround tool, I'd prefer to leave the implementation as is. I would hope that the tool would eventually become obsolete due to more stability in APIs, charts being updated in time or a better solution.
> 
> That is why I don't think it a solution that we will merge at this time.

Nice work. Just compiled and tested your fix and it works well. We only need it for scanning purposes (dry run), so the issue with comments etc doesn't bother us too much at the moment.

**Pull request closed without merging**